mode/jsx/jsx.js

```diff
@@ -0,0 +1,217 @@
+// CodeMirror, copyright (c) by Marijn Haverbeke and others
+// Distributed under an MIT license: http://codemirror.net/LICENSE
+
+(function(mod) {
+  if (typeof exports == "object" && typeof module == "object") // CommonJS
+    mod(require("../../lib/codemirror"), require("../javascript/javascript"),
+        require("../xml/xml"));
+  else if (typeof define == "function" && define.amd) // AMD
+    define(["../../lib/codemirror", "../javascript/javascript", "../xml/xml"],
+           mod);
+  else // Plain browser env
+    mod(CodeMirror);
+})(function(CodeMirror) {
+  "use strict";
+
+  CodeMirror.defineMode("jsx", function(config, parserConfig) {
+    var tagOpen = /^\s*</,
+        closingTagOpen = /^\s*<\//,
+        modes = {
+          'js': CodeMirror.getMode(config, {
+            name: "javascript"
+          }),
+          'xml': CodeMirror.getMode(config, {
+            name: "xml",
+            htmlMode: true,
+            multilineTagIndentFactor: parserConfig.multilineTagIndentFactor,
+            multilineTagIndentPastTag: parserConfig.multilineTagIndentPastTag,
+            allowMissing: true
+          })
+        };
+
+    function tokenBase(stream, state) {
+      if (state.nextMode) {
+        state.currentMode = state.nextMode;
+        state.nextMode = null;
+      }
+
+      // If current mode is not XML and an XML tag opening gets detected next,
+      // switch editor mode to XML.
+      // ATTENTION: This should not happen inside an attribute JavaScript
+      // expression, since XML structures are not valid XML attributes.
+      if (state.currentMode != "xml" && !state.context.inAttrJSExpression) {
+        if (stream.match(tagOpen, false)) {
```

Comment:
You 're right I got this initially wrong because I was trying to be more tolerant over when a tag starts (the XML mode is as well).

I took a look at the syntax of HTML tags (http://www.w3.org/TR/html-markup/syntax.html#syntax-start-tags) and added a stricter control. I am considering a valid "tag open" only when it matches the proper syntax of starting a "Start tag" (`<tagname`) or an "End tag" (`</tagname`) which allows no spaces in between `<`, `/` (if inside an "End tag") and the tag name, which should be a case insensitive alphanumeric.

I think I improved that a little bit (still has some edge cases) with https://github.com/codemirror/CodeMirror/pull/3742/files#diff-8fe55bb9634da263c3fa302cdbefb06cR17.